MVP Agent Implementation

iamjoemccormick opened the pull request:
### What does this PR do / why do we need it?
*Required for all PRs.*
<!--Questions that may be helpful filling out this section:

* What do we gain/lose with this PR?
-->

### Related Issue(s)
*Required when applicable.*
<!-- Link the PR to issues(s) using keywords:
* Reference: https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue
-->

### Where should the reviewer(s) start reviewing this? 
*Only required for larger PRs when this may not be immediately obvious.*
<!-- Questions that may be helpful filling out this section:

* Where should someone start (file/line) to begin reviewing the new/updated functionality in this
  PR? 
* Is there a logical progression the reviewer can follow to navigate their way through the changes
  (i.e., main.go -> api.go -> db.go)?
-->

### Are there any specific topics we should discuss before merging?
*Not required.*
<!-- Questions that may be helpful filling out this section:

* Are there potential tradeoffs in the implementation?
-->

### What are the next steps after this PR? 
*Not required.*
<!-- Questions that may be helpful filling out this section:

* Is further work planned in this area after this PR is merged? 
  * If so, what issue(s) and/or PRs is it tracked in? 
-->

### Checklist before merging:
*Required for all PRs.*

When creating a PR these are items to keep in mind that cannot be checked by GitHub actions:

- [ ] Documentation:
  - [ ] Does developer documentation (code comments, readme, etc.) need to be added or updated?
  - [ ] Does the user documentation need to be expanded or updated for this change?
- [ ] Testing:
  - [ ] Does this functionality require changing or adding new unit tests?
  - [ ] Does this functionality require changing or adding new integration tests?
- [ ] Git Hygiene: 
  - [ ] Do commits adhere to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)?
  - [ ] Is the commit history squashed down reasonably?

For more details refer to the [Go coding standards](https://github.com/ThinkParQ/beegfs-go/wiki/Getting-Started-with-Go#coding-standards) and the [pull request process](https://github.com/ThinkParQ/beegfs-go/wiki/Pull-Requests).

iamjoemccormick committed:
wip: initial agent skeleton

iamjoemccormick committed:
wip: agent config/manifest handling

iamjoemccormick committed:
wip: agent deployer

iamjoemccormick committed:
wip: support multiple fs, simplify deployer+reconciler state

iamjoemccormick committed:
wip: move agent protos to a new package

rustybee42 reviewed (commented):
I'm not very deep into this one, but I wonder why choosing yaml for the manifest  format instead of toml. We already use the latter one and in my view it's superior compared to yaml.

iamjoemccormick commented:
> I'm not very deep into this one, but I wonder why choosing yaml for the manifest format instead of toml. We already use the latter one and in my view it's superior compared to yaml.

I actually started with TOML and intended to support TOML and optionally YAML, but unfortunately the manifest contains a map of file systems, each of which contains maps of agents, each with lists of nodes that are themselves fairly complex structures. It got really awkward to try and express the structure I wanted in TOML so I switched to YAML.

I agree for a reasonably flat config file TOML is simpler to read and reason about. But for modeling deep object hierarchies I think YAML has the edge. This is likely why projects like Ansible and Kubernetes lean heavily on YAML. This manifest is conceptually similar, a runtime description of a full BeeGFS deployment.

I also thought the choice of YAML for the manifest may help distinguish between the static service config defined in TOML files, and dynamic system-wide deployment state in a YAML manifest.

iamjoemccormick committed:
wip: expand/improve common config

- add auth and TLS structs
- reduce boiler plate on common service and source config
- split out filesystem.go into multiple files

iamjoemccormick committed:
wip: rename source to install-source

iamjoemccormick committed:
wip: rename target root_dir to path

iamjoemccormick committed:
wip: rename nodes to services

iamjoemccormick committed:
wip: add generated metadata to manifests

iamjoemccormick committed:
wip: enforce v4 UUIDs and derive a unique short UUID

iamjoemccormick committed:
wip: switch per service install source to a simple executable path

iamjoemccormick committed:
todo: unstage + self-review
